### PR TITLE
Feed sorting buttons

### LINKED ISSUE #172: Allow the feed to be sorted according to the sort buttons
Acceptance criteria:
- general service for sorting posts
- view model for the feed, for the sorting buttons to work

### PR DESCRIPTION
**Pull request by JoachimFavre**
Fixes #172.

This PR adds a service for sorting posts, and uses it in the view-model of the feed to show the posts in the correct order. I have also had to add a very simple view-model for the chosen sorting option.

I think it is interesting for me to justify my design choices for the `PostSortingService`, since there was some debate on whether its sort function should take a general function or an enum. I chose the second, because I felt it allowed a better encapsulation and type safety. I did not find any example of a place where we would want a sorting order different from the ones in the feed. Moreover, in my case, I decided to give a score function as a parameter instead of a comparator. This makes, in my opinion, code more readable since we more clearly see the intended order. However, it requires a boolean to know if we want to sort by increasing or decreasing score. This needs to be bundled together with the score function, and the enum appears to be a good way for doing this as well. I hope you will agree with this design choice.

Furthermore, note that I had to do some modifications to the post and user generators for my testing. I could have done this in another PR, but I think we have already too many PRs for this sprint and that it would just add a blocker for this PR. I therefore decided to put those (very small) modifications in this PR.

Acceptance criteria:
- [x] general service for sorting posts
- [x] view model for the feed, for the sorting buttons to work

### CODE REVIEW

File: test/views/pages/home/dynamic_home_test.dart
Comment (CHOOSEIT) [inline test/views/pages/home/dynamic_home_test.dart:61]: I feel that we can find a better way to get the providerScope and read results from the `feedSortOptionsProvider`. 

We can maybe take inspiration from @yoannLafore's repository testing, where it creates a `ProviderContainer` and reads the provider from it. ([ref](https://github.com/ProximaEPFL/proxima/pull/181/files#diff-9f98d592c3821b134d5ac116b3644a38358e76cd72a4b974d6468c572e7a04e3)).

I don't know if this the best way, but I found a way to make this idea working:
I create the homeWidget with: 
```dart 
homepageWidget = UncontrolledProviderScope(
  container: containerProvider,
  child: homePageApp,
);

 ```
where 
```dart
containerProvider = ProviderContainer(overrides: [
  firestoreProvider.overrideWithValue(firestore),
  geoLocationServiceProvider.overrideWithValue(geoLocationService),
]);

 ```
 
Using that, I am able to use: 
```dart 
expect(
  containerProvider.read(feedSortOptionsProvider),
  equals(sortOption),
);

// Expect the list to be in correct order
// We compare the post id values for an easier debugging.
final actualPostsOverview = await containerProvider.read(
  postOverviewProvider.future,
);


```

But since we never used `UncontrolledProviderScope` in our codebase, I am not sure if this is a better solution.

**Review by CHOOSEIT — COMMENTED** [review]
Really clean. Great work 🎉.

I have a comment about a test.

However, overall, I really love the way you test your code, creating loops and functions to avoid code duplication. 💯

File: lib/services/sorting/post_sorting_service.dart
Comment (yoannLafore) [inline lib/services/sorting/post_sorting_service.dart:21]: I just believe we can specify the parameter types in the function here and maybe rename them.
```suggestion
    int defaultComparator(PostFirestore postA, PostFirestore postB) {
```

File: lib/views/home_content/feed/feed_sort_option_chips.dart
Comment (yoannLafore) [inline lib/views/home_content/feed/feed_sort_option_chips.dart:8]: Very small suggestion, I think we can make this variable to the plurial given that it is a map with multiple keys.
```suggestion
  static final optionChipKeys = Map.fromIterables(
```

File: lib/views/home_content/feed/feed_sort_option_chips.dart
Comment (yoannLafore) [inline lib/views/home_content/feed/feed_sort_option_chips.dart:34]: Good job with the UI :+1: 
Small suggestion, I believe that we can use the arrow synthax here.

```suggestion
        final sortOptions = PostSortOption.values
        .map(
          (sortOption) => ChoiceChip(
            key: optionChipKey[sortOption],
            selected: sortOption == selectedOption,
            label: Text(sortOption.name),
            onSelected: (value) {
              if (value) {
                ref
                    .read(feedSortOptionsProvider.notifier)
                    .setSortOption(sortOption);
              }
            },
          ),
        )
        .toList();
```

File: test/services/sorting/post_sorting_service_test.dart
Comment (yoannLafore) [inline test/services/sorting/post_sorting_service_test.dart:21]: Very nice job with the testing ! 
I have little suggestion, I think we should declare those kind of variables as `late` that we initalize in the `setUp` function. I believe this is the convention we adopted for the other tests and it ensures that each test has those variables freshly created when running and so that they were not modified by other tests.

I also wonder if we could add a "global" group to encapsulate all the testing.
I think we could have something like:

```suggestion
  group("Testing post sorting service", () {
    late PostSortingService sortingService;

    late FirestorePostGenerator postGenerator;
    late List<GeoPoint> positions;
    late List<PostFirestore> posts;

    setUp(() {
      sortingService = PostSortingService();

      postGenerator = FirestorePostGenerator();
      positions = GeoPointGenerator.generatePositions(userPosition0, 10, 0);
      posts = postGenerator.generatePostsAtDifferentLocations(positions);
    });
    
    // All the other groups that are performing tests
    // ...
    
  }
```

Note: if you go with this suggestion, I think you'll need to also implement the next suggestion otherwise I believe you'll get an error because some variable won't be initialized.

File: test/services/sorting/post_sorting_service_test.dart
Comment (yoannLafore) [inline test/services/sorting/post_sorting_service_test.dart:139]: For similar reasons to my previous comment, I think we could make those `late` and initialize them in the `setUp`.
```suggestion
      late List<GeoPoint> positionsToPutOnTop;
      late List<PostFirestore> postsToPutOnTop;
      late List<PostFirestore> existingPostsToPutOnTop;

      setUp(() {
        positionsToPutOnTop =
            GeoPointGenerator.generatePositions(userPosition0, 5, 0);

        postsToPutOnTop = postGenerator
            .generatePostsAtDifferentLocations(positionsToPutOnTop);

        existingPostsToPutOnTop = postsToPutOnTop.take(3).toList();
      });
```

**Review by yoannLafore — COMMENTED** [review]
You did a very good job at implementing the sorting service !
The code is very clean and robustly tested.

I also like how you made use of functions in your tests to reduce code duplication.
As @CHOOSEIT said, I also just think that it could have a little more documentation :+1: 

Overall, the PR looks good to me.
I have just made a few suggestions below.

Thank you for you work !
Great job :tada:

File: lib/services/sorting/post_sort_option.dart
Comment (JoachimFavre) [reply lib/services/sorting/post_sort_option.dart:40]: As I mention in my code, the absolute value of those score does not matter, only the relative values. This means that subtracting currentTime from every score does not change anything at best, and, at worst, makes it non-deterministic: `currentTime` will be evaluated every time we want to compare two posts, so, for two posts a,b published at the same time, we might have both a < b and b < a if it takes more than a millisecond to compute the score of the first one one.
Would you want me to increase  this documentation?

File: test/views/pages/home/dynamic_home_test.dart
Comment (JoachimFavre) [reply test/views/pages/home/dynamic_home_test.dart:61]: By humble curiosity, why do you call this a "better way"? This looks like more boilerplate to me at first sight.

File: test/services/sorting/post_sorting_service_test.dart
Comment (JoachimFavre) [reply test/services/sorting/post_sorting_service_test.dart:21]: I have a question and a remark on this suggestion:
- Does this really make sense to put all my tests in the same group? This really feels like using groups just for using groups, without having them giving real quality to me at first sight.
- My idea was that it was a lot of boilerplate to declare my variables as `late` and define them in the setup; since they never change. However, on reflection, this appears to be a good way to avoid future mistakes and non-deterministic tests. I'll implement this.

File: test/views/pages/home/dynamic_home_test.dart
Comment (CHOOSEIT) [reply test/views/pages/home/dynamic_home_test.dart:61]: I feel like we are extracting the container from the UI, when we could have extracted it more globally without going through the UI, but I may be wrong.

File: test/services/sorting/post_sorting_service_test.dart
Comment (yoannLafore) [reply test/services/sorting/post_sorting_service_test.dart:21]: - I don't know if it is strictly necessary but I quite like the idea having a big group that encapsulates everything and that tells what service we are testing. And then we have smaller groups to tell which part of the service we are testing. But this is only a personal opinion.
- Nice thank you :+1:

File: lib/services/sorting/post_sort_option.dart
Comment (yoannLafore) [reply lib/services/sorting/post_sort_option.dart:40]: Oh ok I see thank you.
Adding a bit more documentation could be a good idea you're right :+1:

File: test/views/pages/home/dynamic_home_test.dart
Comment (JoachimFavre) [reply test/views/pages/home/dynamic_home_test.dart:61]: What I did is what is recommended [in the documentation](https://riverpod.dev/docs/essentials/testing) for testing both widgets and providers at the same time. In my opinion, this is ok.

File: lib/services/sorting/post_sorting_service.dart
Comment (JoachimFavre) [reply lib/services/sorting/post_sorting_service.dart:14]: Well seen for the typo, fixed in 0f0e6f26c6958427a4f3032b1d8ae705846e37ef.

No `K` is not a typo, if an element is in `putOnTop` but not in the list, it will not be added to the list. This is for challenges later so that, if they are in the feed, they appear on top. I improved the documentation in c1909e7af60f5f2a37443cf91e365646da5052cb. Tell me if this is clearer.

File: lib/services/sorting/post_sorting_service.dart
Comment (JoachimFavre) [reply lib/services/sorting/post_sorting_service.dart:57]: Fixed in 4bc9f8a131b5635dbc7e72269dcdf31ce85847da.

File: lib/views/home_content/feed/feed_sort_option_chips.dart
Comment (JoachimFavre) [reply lib/views/home_content/feed/feed_sort_option_chips.dart:27]: Fixed in 9d7384f7bcf8a380745db7b144ff99131f04d62c.

File: test/services/sorting/post_sorting_service_test.dart
Comment (JoachimFavre) [reply test/services/sorting/post_sorting_service_test.dart:109]: Fix in bb5caded1305cb7d8977074298e655dc025ffacb. Tell me if this is clearer.

File: lib/services/sorting/post_sort_option.dart
Comment (JoachimFavre) [reply lib/services/sorting/post_sort_option.dart]: Fixed in 3f486895e8f6c6ad7f78b3bb60e3029aa494b14e.

File: lib/services/sorting/post_sort_option.dart
Comment (JoachimFavre) [reply lib/services/sorting/post_sort_option.dart]: Indeed, some documentation was missing, I added it in 024c9bfc9b5d57e0fb1e6395b39f682e76a5c3c8.

In fact, writing documentation made me realise I had confused myself when writing this code and its corresponding test. This was fixed in d1ad77617eab46c9cdbad9f173d9cf632f321f56.

File: test/views/pages/home/dynamic_home_test.dart
Comment (CHOOSEIT) [reply test/views/pages/home/dynamic_home_test.dart:61]: You are right. Thank you.

File: lib/services/sorting/post_sorting_service.dart
Comment (JoachimFavre) [reply lib/services/sorting/post_sorting_service.dart]: Fixed in d36c080e5f63c82ff79a2853138390e7856148d8.

File: lib/views/home_content/feed/feed_sort_option_chips.dart
Comment (JoachimFavre) [reply lib/views/home_content/feed/feed_sort_option_chips.dart]: Fixed in f5420ecd18367f611c0e3939ef52571ab692c501.

File: lib/views/home_content/feed/feed_sort_option_chips.dart
Comment (JoachimFavre) [reply lib/views/home_content/feed/feed_sort_option_chips.dart]: Fixed in b0dec4962087d323e1efed7f1238fd02a8cf7235.

File: test/services/sorting/post_sorting_service_test.dart
Comment (JoachimFavre) [reply test/services/sorting/post_sorting_service_test.dart]: - If you're ok with this, I will not implement this suggestion. I think test names are already long enough, and that this is not strictly necessary.
- Fixed in d6c9c4ecd56bf2b95cdcdb3926ee37184c6624b8.

File: test/services/sorting/post_sorting_service_test.dart
Comment (JoachimFavre) [reply test/services/sorting/post_sorting_service_test.dart]: Fixed in d6c9c4ecd56bf2b95cdcdb3926ee37184c6624b8.

File: test/services/sorting/post_sorting_service_test.dart
Comment (CHOOSEIT) [reply test/services/sorting/post_sorting_service_test.dart]: Much clearer; thank you !

File: test/services/sorting/post_sorting_service_test.dart
Comment (yoannLafore) [reply test/services/sorting/post_sorting_service_test.dart]: I understand, it's alright for me :+1:

File: test/services/sorting/post_sorting_service_test.dart
Comment (CHOOSEIT) [reply test/services/sorting/post_sorting_service_test.dart]: Much clearer for the `testOption` function, thank you. However, I think that the inner mechanism of the function needs a little more explanation. Like, why `referenceScore` and what it represents, because this function is really abstract. :smile

Comment (JoachimFavre) [reply test/services/sorting/post_sorting_service_test.dart]: Oh, I understand better. Is 1ffcb65fdd81fcb829228ecd467bf2591ba02662 good for you?

**Review by yoannLafore — APPROVED** [review]
Thank you for bringing the modifications !
The PR looks good to me now.
Amazing job :tada:

File: test/services/sorting/post_sorting_service_test.dart
Comment (CHOOSEIT) [reply test/services/sorting/post_sorting_service_test.dart]: Much clearer 💯 . Thank you.

**Review by CHOOSEIT — APPROVED** [review]
LGTM now ! 🎉 

Great work 💯

**Comment by JoachimFavre** [comment]
I found some mistake in my code. I'll let @yoannLafore re-review before I merge.

**Review by yoannLafore — APPROVED** [review]
Good catch !
Thank you for the change, this looks good to me :+1:

**Pull request merged**
